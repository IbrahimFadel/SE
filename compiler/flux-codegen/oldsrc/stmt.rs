// use super::*;

// impl<'a> Codegen<'a> {
// 	pub fn block_stmt(&mut self, block: &BlockStmt) -> Option<FluxError> {
// 		for stmt in block {
// 			if let Some(err) = self.stmt(stmt) {
// 				return Some(err);
// 			}
// 		}
// 		return None;
// 	}

// 	fn stmt(&mut self, stmt: &Stmt) -> Option<FluxError> {
// 		match stmt {
// 			Stmt::If(if_) => self.if_stmt(if_),
// 			Stmt::VarDecl(var) => self.var_decl(var),
// 			Stmt::Return(ret) => self.ret_stmt(ret),
// 			Stmt::ExprStmt(expr) => self.expr(expr).1,
// 			_ => Some(self.error(
// 				format!(
// 					"internal compiler error: could not codegen statement: {:?}",
// 					stmt
// 				),
// 				FluxErrorCode::CodegenCouldNotCodegenStmt,
// 				vec![],
// 			)),
// 		}
// 	}

// 	fn if_stmt(&mut self, if_: &If) -> Option<FluxError> {
// 		let (cond, err) = self.expr(&if_.condition);
// 		if let Some(err) = err {
// 			return Some(err);
// 		}

// 		unsafe {
// 			let f = self
// 				.tmp_data
// 				.cur_fn
// 				.expect("expected to be in function: internal compiler error");
// 			let then = LLVMAppendBasicBlock(f, str_to_cstring("then"));
// 			let else_ = LLVMCreateBasicBlockInContext(self.ctx, str_to_cstring("else"));
// 			let mut cont: LLVMBasicBlockRef = ptr::null_mut();
// 			if if_.else_.is_some() {
// 				cont = LLVMCreateBasicBlockInContext(self.ctx, str_to_cstring("continue"));
// 				self.tmp_data.merging_bb = Some(cont);
// 			} else {
// 				self.tmp_data.merging_bb = Some(else_);
// 			}

// 			LLVMBuildCondBr(self.builder, cond, then, else_);

// 			self.tmp_data.cur_bb = Some(then);
// 			LLVMPositionBuilderAtEnd(self.builder, then);

// 			let name = get_basicblock_name(then);
// 			let initial_scope = self.symbol_table.cur_scope.clone();
// 			self
// 				.symbol_table
// 				.scope_conns
// 				.push((initial_scope.clone(), name.clone()));
// 			self
// 				.symbol_table
// 				.scope_values
// 				.insert(name.clone(), HashMap::new());
// 			self.symbol_table.cur_scope = name;

// 			if let Some(err) = self.block_stmt(&if_.then) {
// 				return Some(err);
// 			}
// 			if if_.else_.is_some() {
// 				LLVMBuildBr(self.builder, cont);
// 			} else {
// 				LLVMBuildBr(self.builder, else_);
// 			}

// 			self.tmp_data.cur_bb = Some(else_);
// 			LLVMPositionBuilderAtEnd(self.builder, else_);

// 			if let Some(else_block) = &if_.else_ {
// 				let name = get_basicblock_name(else_);
// 				self
// 					.symbol_table
// 					.scope_conns
// 					.push((initial_scope, name.clone()));
// 				self
// 					.symbol_table
// 					.scope_values
// 					.insert(name.clone(), HashMap::new());
// 				self.symbol_table.cur_scope = name;

// 				LLVMAppendExistingBasicBlock(f, else_);
// 				self.block_stmt(else_block);

// 				LLVMBuildBr(self.builder, cont);
// 				LLVMAppendExistingBasicBlock(f, cont);
// 				self.tmp_data.cur_bb = Some(cont);
// 				LLVMPositionBuilderAtEnd(self.builder, cont);
// 			} else {
// 				LLVMAppendExistingBasicBlock(f, else_);
// 				LLVMPositionBuilderAtEnd(self.builder, else_);
// 			}
// 		}

// 		return None;
// 	}

// 	fn ret_stmt(&mut self, ret: &Return) -> Option<FluxError> {
// 		unsafe {
// 			match &ret.val {
// 				Some(x) => {
// 					let (e, err) = self.expr(&x);
// 					if let Some(err) = err {
// 						return Some(err);
// 					}
// 					LLVMBuildRet(self.builder, e);
// 					return None;
// 				}
// 				None => {
// 					LLVMBuildRetVoid(self.builder);
// 					return None;
// 				}
// 			};
// 		}
// 	}
// }
