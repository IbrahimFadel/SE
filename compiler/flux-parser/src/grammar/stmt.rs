use flux_lexer::T;

use super::{expr::type_expr, *};

pub(crate) fn stmt(p: &mut Parser) -> CompletedMarker {
	if p.at(T!(let))
		|| p.at(T!(iN))
		|| p.at(T!(uN))
		|| p.at(T!(f32))
		|| p.at(T!(f64))
		|| (p.at(T!(ident)) && matches!(p.peek_next(), Some(T!(ident)) | Some(T!(eq))))
		|| p.at(T!(lparen))
	{
		var_decl(p)
	} else if p.at(T!(return)) {
		return_stmt(p)
	} else {
		let m = p.start();
		expr::expr(p, true);
		if p.at(T!(semicolon)) {
			p.bump();
		}
		m.complete(p, SyntaxKind::ExprStmt)
	}
}

fn return_stmt(p: &mut Parser) -> CompletedMarker {
	let m = p.start();
	p.bump();
	if p.at(T!(semicolon)) {
		p.bump();
		return m.complete(p, SyntaxKind::ReturnStmt);
	}
	expr::expr(p, true);
	p.expect(
		T!(semicolon),
		format!("expected `;` after return statement"),
	);
	m.complete(p, SyntaxKind::ReturnStmt)
}

// fn else_if_stmt(p: &mut Parser) -> CompletedMarker {
// 	let m = p.start();
// 	p.expect(T!(if), format!("expected `if` in else if statement"));
// 	expr::expr(p, false);
// 	block(p);
// 	m.complete(p, SyntaxKind::IfStmt)
// }

fn var_decl(p: &mut Parser) -> CompletedMarker {
	let m = p.start();
	if p.at(T!(let)) {
		p.bump();
	} else {
		type_expr(p);
	}
	p.expect(
		TokenKind::Ident,
		format!("expected identifier in variable declaration"),
	);
	p.expect(
		TokenKind::Eq,
		format!("expected `=` in variable declaration"),
	);
	expr::expr(p, true);
	p.expect(
		TokenKind::SemiColon,
		format!("expected `;` after variable declaration"),
	);
	m.complete(p, SyntaxKind::VarDecl)
}

#[cfg(test)]
mod tests {
	use crate::test_stmt_str;
	test_stmt_str!(var_decl, "i32 x = 1;");
	test_stmt_str!(
		var_decl_fcont,
		r#"i32 y =
	i32 x = 1;"#
	);
}
