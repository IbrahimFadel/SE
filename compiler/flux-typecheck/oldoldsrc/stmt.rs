// use flux_ast::{Return, VarDecl};

// use super::*;

// impl<'a> TypecheckCtx<'a> {
// 	pub fn check_stmt(&mut self, stmt: &'a mut Stmt) -> Option<FluxError> {
// 		match stmt {
// 			Stmt::VarDecl(var) => self.check_var(var),
// 			Stmt::Return(ret) => self.check_ret(ret),
// 			Stmt::ExprStmt(expr) => self.check_expr(expr),
// 			_ => None,
// 		}
// 	}

// 	fn check_ret(&mut self, ret: &'a mut Return) -> Option<FluxError> {
// 		if let Some(x) = &mut ret.val {
// 			if let Some(err) = self.check_expr(x) {
// 				return Some(err);
// 			}
// 		}
// 		return None;
// 	}

// 	fn check_var(&mut self, var: &'a mut VarDecl) -> Option<FluxError> {
// 		self.expecting_ty = Some(&var.type_);
// 		for name in &var.names {
// 			self.var_types.insert(name.to_string(), &var.type_);
// 		}

// 		for val in &mut var.values {
// 			if let Some(err) = self.check_expr(val) {
// 				return Some(err);
// 			}
// 		}

// 		return None;
// 	}
// }
