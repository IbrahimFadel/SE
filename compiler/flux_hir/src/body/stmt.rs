use super::*;

impl<'a, 'pkgs> LowerCtx<'a, 'pkgs> {
    pub(crate) fn lower_stmt(
        &mut self,
        stmt: ast::Stmt,
        generic_params: &GenericParams,
    ) -> (bool, Typed<ExprIdx>) {
        match stmt {
            ast::Stmt::LetStmt(let_stmt) => (false, self.lower_let_expr(let_stmt, generic_params)),
            ast::Stmt::ExprStmt(expr_stmt) => {
                (false, self.lower_expr_stmt(expr_stmt, generic_params))
            }
            ast::Stmt::TerminatorExprStmt(terminator_expr_stmt) => (
                true,
                self.lower_terminator_expr_stmt(terminator_expr_stmt, generic_params),
            ),
        }
    }

    // Always return unit `()`
    fn lower_let_expr(
        &mut self,
        let_stmt: ast::LetStmt,
        generic_params: &GenericParams,
    ) -> Typed<ExprIdx> {
        let name = self.lower_name(let_stmt.name());
        let ty = let_stmt
            .ty()
            .map(|ty| self.lower_type(Some(ty), generic_params).inner)
            .unwrap_or_else(|| {
                self.tckh
                    .tenv
                    .insert(Type::Unknown.file_span(self.file_id, name.span))
            });
        let val = self.lower_expr(let_stmt.value(), generic_params);
        let unification_span = self.tckh.tenv.get_filespan(&val.tid);
        self.tckh
            .unify(ty, val.tid, unification_span)
            .unwrap_or_else(|err| self.diagnostics.push(err));

        let tid = let_stmt.ty().map_or(val.tid, |_| ty);

        self.tckh.tenv.insert_local(name.inner, tid);
        self.alloc_expr(Expr::Tuple(vec![])).with_type(val.tid)
    }

    fn lower_expr_stmt(
        &mut self,
        expr_stmt: ast::ExprStmt,
        generic_params: &GenericParams,
    ) -> Typed<ExprIdx> {
        self.lower_expr(expr_stmt.expr(), generic_params)
    }

    fn lower_terminator_expr_stmt(
        &mut self,
        terminator_expr_stmt: ast::TerminatorExprStmt,
        generic_params: &GenericParams,
    ) -> Typed<ExprIdx> {
        self.lower_expr(terminator_expr_stmt.expr(), generic_params)
    }
}
