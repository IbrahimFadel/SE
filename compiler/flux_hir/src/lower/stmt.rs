use crate::hir::{LetStmt, Stmt};

use super::*;

/// The lowered [`Stmt`], whether or not the [`ast::Stmt`] had a semicolon (needed for
/// typechecking), and its type
type StmtResult = (Stmt, bool, TypeId);

impl LoweringCtx {
    pub(crate) fn lower_stmt(&mut self, stmt: ast::Stmt) -> StmtResult {
        match stmt {
            ast::Stmt::LetStmt(let_stmt) => self.lower_let_stmt(let_stmt),
            ast::Stmt::ExprStmt(expr) => self.lower_expr_stmt(expr),
        }
    }

    fn lower_let_stmt(&mut self, let_stmt: ast::LetStmt) -> StmtResult {
        let name = self.lower_node(
            let_stmt.name(),
            |this, _| {
                Spanned::new(
                    this.interner.get_or_intern_static(POISONED_STRING_VALUE),
                    this.span_node(&let_stmt),
                )
            },
            |this, name| Spanned::new(name.ident().unwrap().text_key(), this.span_node(&let_stmt)),
        );

        let ty = if let Some(ty) = let_stmt.ty() {
            let ty = self.lower_type(ty);
            self.tchk.tenv.insert(self.file_spanned(self.to_ts_ty(&ty)))
        } else {
            self.tchk.tenv.insert(self.file_spanned(Spanned::new(
                ts::Type::new(TypeKind::Unknown),
                self.span_node(&let_stmt),
            )))
        };

        let (value, value_ty_id) = self.lower_node(
            let_stmt.value(),
            |this, _| {
                (
                    this.exprs
                        .alloc(Spanned::new(Expr::Error, this.span_node(&let_stmt))),
                    this.tchk.tenv.insert(this.file_spanned(Spanned::new(
                        ts::Type::new(TypeKind::Unknown),
                        this.span_node(&let_stmt),
                    ))),
                )
            },
            |this, expr| this.lower_expr(expr),
        );

        let result = self
            .tchk
            .unify(ty, value_ty_id, self.file_span(self.span_node(&let_stmt)));
        self.maybe_emit_diagnostic(result);

        self.tchk.tenv.insert_local_to_scope(name.inner, ty);

        (Stmt::LetStmt(LetStmt { name, ty, value }), true, ty)
    }

    fn lower_expr_stmt(&mut self, expr: ast::ExprStmt) -> StmtResult {
        let (idx, ty_id) = self.lower_node(
            expr.expr(),
            |this, _| {
                let expr = Spanned::new(Expr::Error, this.span_node(&expr));
                let ty_id = this.tchk.tenv.insert(
                    this.file_spanned(Spanned::new(ts::Type::new(TypeKind::Unknown), expr.span)),
                );
                let idx = this.exprs.alloc(expr);
                (idx, ty_id)
            },
            |this, expr| this.lower_expr(expr),
        );
        let has_semicolon = expr.semicolon().is_some();
        (Stmt::ExprStmt(idx), has_semicolon, ty_id)
    }
}
