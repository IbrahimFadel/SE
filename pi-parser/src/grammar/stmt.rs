use pi_lexer::T;

use super::{expr::type_expr, *};

pub(crate) fn block(p: &mut Parser) -> CompletedMarker {
	let m = p.start();
	p.expect(T![lbrace], format!("expected `{{` at start of block"));
	while !p.at(T![rbrace]) && !p.at_end() {
		stmt(p);
	}
	p.expect(T![rbrace], format!("expected `}}` at end of block"));
	m.complete(p, SyntaxKind::BlockStmt)
}

pub(crate) fn stmt(p: &mut Parser) -> Option<CompletedMarker> {
	if p.at(T!(iN)) {
		Some(var_decl(p))
	} else if p.at(T!(if)) {
		Some(if_stmt(p))
	} else {
		expr::expr(p)
	}
}

fn if_stmt(p: &mut Parser) -> CompletedMarker {
	let m = p.start();
	p.expect(T!(if), format!("expected `if` in if statement"));
	expr::expr(p);
	block(p);
	while p.at(T!(else)) {
		p.bump();
		if p.at(T!(lbrace)) {
			block(p);
		} else {
			else_if_stmt(p);
		}
	}
	m.complete(p, SyntaxKind::IfStmt)
}

fn else_if_stmt(p: &mut Parser) -> CompletedMarker {
	let m = p.start();
	p.expect(T!(if), format!("expected `if` in else if statement"));
	expr::expr(p);
	block(p);
	m.complete(p, SyntaxKind::IfStmt)
}

fn var_decl(p: &mut Parser) -> CompletedMarker {
	let m = p.start();
	type_expr(p);
	p.expect(
		TokenKind::Ident,
		format!("expected identifier in variable declaration"),
	);
	p.expect(
		TokenKind::Eq,
		format!("expected `=` in variable declaration"),
	);
	expr::expr(p);
	p.expect(
		TokenKind::SemiColon,
		format!("expected `;` after variable declaration"),
	);
	m.complete(p, SyntaxKind::VarDecl)
}

#[cfg(test)]
mod tests {
	use crate::test_stmt_str;
	test_stmt_str!(var_decl, "i32 x = 1;");
	test_stmt_str!(
		var_decl_fcont,
		r#"i32 y =
	i32 x = 1;"#
	);
}
