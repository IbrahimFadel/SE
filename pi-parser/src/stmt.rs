use pi_ast::{BlockStmt, Expr, If, Mod, Return, Stmt, VarDecl};
use pi_error::PIErrorCode;
use pi_lexer::token::TokenKind;

use super::Parser;

impl<'a> Parser<'a> {
	pub fn mod_stmt(&mut self, pub_: bool) -> Mod {
		self.next();
		let name = self.ident();
		self.expect(
			TokenKind::Semicolon,
			self.error(
				"expected `;` after mod statement".to_owned(),
				PIErrorCode::ParseExpectedSemicolonAfterModStmt,
				vec![],
			),
		);
		if self.tok().kind == TokenKind::Semicolon {
			self.next();
		}
		Mod::new(pub_, name)
	}

	pub fn block(&mut self) -> BlockStmt {
		self.expect(
			TokenKind::LBrace,
			self.error(
				"expected `{` at beginning of block".to_owned(),
				PIErrorCode::ParseExpectedLBraceInBlock,
				vec![(
					format!(
						"expected `{{` at beginning of block, instead got {}",
						Parser::tok_val(self.program, self.tok())
					),
					self.tok().span.clone(),
				)],
			),
		);
		if self.tok().kind == TokenKind::LBrace {
			self.next();
		}

		let mut stmts = vec![];
		while self.tok().kind != TokenKind::RBrace && self.tok().kind != TokenKind::EOF {
			let stmt = self.stmt();
			if stmt == Stmt::Error {
				break;
			}
			stmts.push(stmt);
		}

		self.expect(
			TokenKind::RBrace,
			self.error(
				"expected `}` at end of block".to_owned(),
				PIErrorCode::ParseExpectedRBraceInBlock,
				vec![(
					format!(
						"expected `}}` at end of block, instead got {}",
						Parser::tok_val(self.program, self.tok())
					),
					self.tok().span.clone(),
				)],
			),
		);
		if self.tok().kind == TokenKind::RBrace {
			self.next();
		}

		return stmts;
	}

	fn stmt(&mut self) -> Stmt {
		if self.tok().kind > TokenKind::TypesBegin && self.tok().kind < TokenKind::TypesEnd {
			return self.var_decl_stmt();
		} else if self.tok().kind == TokenKind::Return {
			return self.return_stmt();
		} else if self.tok().kind == TokenKind::If {
			return self.if_stmt();
		} else {
			let expr = self.expr();
			self.expect(
				TokenKind::Semicolon,
				self.error(
					"expected `;` after expression".to_owned(),
					PIErrorCode::ParseExpectedSemicolonAfterExpr,
					vec![],
				),
			);
			if self.tok().kind == TokenKind::Semicolon {
				self.next();
			}
			if expr == Expr::Error {
				return Stmt::Error;
			}
			return Stmt::ExprStmt(expr);
		}
	}

	fn if_stmt(&mut self) -> Stmt {
		self.next();
		let condition = self.expr();
		let then = self.block();
		let mut else_ = None;
		if self.tok().kind == TokenKind::Else {
			self.next();
			if self.tok().kind == TokenKind::If {
				let if_ = self.if_stmt();
				else_ = Some(vec![if_]);
			} else {
				else_ = Some(self.block());
			}
		}
		Stmt::If(If::new(Box::from(condition), then, else_))
	}

	fn return_stmt(&mut self) -> Stmt {
		self.next();
		if self.tok().kind == TokenKind::Semicolon {
			self.next();
			return Stmt::Return(Return::new(None));
		}
		let expr = self.expr();
		self.expect(
			TokenKind::Semicolon,
			self.error(
				"expected `;` after return statement".to_owned(),
				PIErrorCode::ParseExpectedSemicolonAfterReturnStmt,
				vec![],
			),
		);
		if self.tok().kind == TokenKind::Semicolon {
			self.next();
		}
		Stmt::Return(Return::new(Some(expr)))
	}

	fn var_decl_stmt(&mut self) -> Stmt {
		let ty = self.type_expr();

		if self.tok().kind != TokenKind::Ident {
			self.errors.push(self.error(
				"expected identifier in variable declaration".to_owned(),
				PIErrorCode::ParseExpectedIdentVarDecl,
				vec![(
					format!(
						"expected identifier in variable declaration, instead got `{}`",
						Parser::tok_val(self.program, self.tok())
					),
					self.tok().span.clone(),
				)],
			));
		}

		let names_start = self.tok().span.start;
		let mut names_end = self.tok().span.end;
		let mut names = vec![];
		while self.tok().kind != TokenKind::Eq && self.tok().kind != TokenKind::Semicolon {
			names_end = self.tok().span.clone().end;
			let name = self.ident();
			names.push(name);
			if self.tok().kind != TokenKind::Eq && self.tok().kind != TokenKind::Semicolon {
				self.expect(
					TokenKind::Comma,
					self.error(
						"expected `,` in variable declaration identifier list".to_owned(),
						PIErrorCode::ParseExpectedCommaInVarDeclIdentList,
						vec![(
							format!(
								"expected `,` in variable declaration identifier list, instead got `{}`",
								Parser::tok_val(self.program, self.tok())
							),
							self.tok().span.clone(),
						)],
					),
				);
				if self.tok().kind == TokenKind::Comma {
					self.next();
				}
			}
		}

		if self.tok().kind == TokenKind::Semicolon {
			self.next();
			return Stmt::VarDecl(VarDecl::new(ty, names, None));
		}
		self.expect(
				TokenKind::Eq,
				self.error(
					"expected either `;` or `=` after variable declaration identifier list".to_owned(),
					PIErrorCode::ParseExpectedSemicolonEqVarDeclIdentList,
					vec![
						(format!("expected either `;` or `=` after variable declaration identifier list, instead got `{}`", Parser::tok_val(self.program, self.tok())), self.tok().span.clone())
					],
				),
			);
		self.next();

		let values_start = self.tok().span.start;
		let mut values_end = self.tok().span.end;
		let mut values = vec![];
		while self.tok().kind != TokenKind::Semicolon {
			values_end = self.tok().span.end;
			let val = self.expr();
			values.push(val);
			if self.tok().kind == TokenKind::Comma {
				self.next();
				if self.tok().kind == TokenKind::Semicolon {
					self.errors.push(self.error(
						"expected expression after comma in variable declaration value list".to_owned(),
						PIErrorCode::ParseExpectedExprAfterCommaVarDeclValueList,
						vec![(
							format!(
								"expected expression after comma in variable declaration value list, instead got `{}`",
								Parser::tok_val(self.program, self.tok())
							),
							self.tok().span.clone(),
						)],
					));
					break;
				}
			} else if self.tok().kind != TokenKind::Semicolon {
				self.errors.push(self.error(
					"expected `;` after variable declaration".to_owned(),
					PIErrorCode::ParseExpectedSemicolonAfterVarDecl,
					vec![(
						format!(
							"expected `;` after variable declaration, instead got `{}`",
							Parser::tok_val(self.program, self.tok())
						),
						self.tok().span.clone(),
					)],
				));
				break;
			}
		}
		if self.tok().kind == TokenKind::Semicolon {
			self.next();
		}

		if names.len() > 1 && values.len() > 1 && values.len() != names.len() {
			if names.len() > values.len() {
				self.errors.push(self.error("more variables than values in variable declaration".to_owned(), PIErrorCode::ParseMoreIdentsThanValsVarDecl, vec![
				(format!("more variables than values in variable declaration: {} values assigned to {} variables", values.len(), names.len()), names_start..self.tok().span.end),
				(format!("{} variables declared", names.len()), names_start..names_end),
				(format!("{} values assigned", values.len()), values_start..values_end),
				("(hint) you can assign one value to multiple variables".to_owned(), names_start..self.tok().span.end),
			]));
			} else {
				self.errors.push(self.error("more values than variables in variable declaration".to_owned(), PIErrorCode::ParseMoreValsThanIdentsVarDecl, vec![
				(format!("more values than variables in variable declaration: {} values assigned to {} variables", values.len(), names.len()), names_start..self.tok().span.end),
				(format!("{} variables declared", names.len()), names_start..names_end),
				(format!("{} values assigned", values.len()), values_start..values_end),
			]));
			}
		}

		return Stmt::VarDecl(VarDecl::new(ty, names, Some(values)));
	}
}
