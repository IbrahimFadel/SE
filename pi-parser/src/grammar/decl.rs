use super::{expr::type_expr, *};
use crate::grammar::stmt::block;
use pi_lexer::T;

pub(crate) fn top_level_decl(p: &mut Parser) {
	if p.at(T!(fn)) {
		fn_decl(p);
	} else {
		p.error(format!("expected top level declaration"));
	}
}

fn fn_decl(p: &mut Parser) -> CompletedMarker {
	assert!(p.at(T![fn]));
	let m = p.start();
	p.bump();
	p.expect(
		T![ident],
		format!("expected identifier in function declaration"),
	);
	if p.at(T![cmplt]) {
		generic_list(p);
	}
	fn_params(p);
	if p.at(T![arrow]) {
		p.bump();
		type_expr(p);
	}
	block(p);
	m.complete(p, SyntaxKind::FnDecl)
}

fn generic_list(p: &mut Parser) -> CompletedMarker {
	assert!(p.at(T![cmplt]));
	let m = p.start();
	p.expect(
		T![ident],
		format!("expected identifier in function generic list"),
	);
	m.complete(p, SyntaxKind::GenericList)
}

fn fn_params(p: &mut Parser) -> CompletedMarker {
	let m = p.start();
	p.expect(
		T![lparen],
		format!("expected `(` at beginning of function parameter list"),
	);
	while !p.at(T![rparen]) && !p.at_end() {
		fn_param(p);
		if !p.at(T![rparen]) {
			if !p.at(T![comma]) {
				p.error(format!(
					"expected either `,` or `)` in function parameter list"
				));
				break;
			} else {
				p.bump();
			}
		}
	}
	p.expect(
		T![rparen],
		format!("expected `)` at end of function parameter list"),
	);
	m.complete(p, SyntaxKind::FnParams)
}

fn fn_param(p: &mut Parser) -> CompletedMarker {
	let m = p.start();
	if p.at(T![mut]) {
		p.bump();
	}
	type_expr(p);
	p.expect(
		T![ident],
		format!("expected identifier in function parameter"),
	);
	m.complete(p, SyntaxKind::FnParam)
}

#[cfg(test)]
mod test {
	use crate::test_decl_str;

	test_decl_str!(fn_decl, r#"fn main(i32 argc, mut i32 test) -> i32 {}"#);
}
