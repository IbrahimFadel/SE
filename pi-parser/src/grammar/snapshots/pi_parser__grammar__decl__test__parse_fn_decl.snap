---
source: pi-parser/src/grammar/decl.rs
expression: "&parse.debug_tree()"
---
Root@0..41
  FnDecl@0..41
    FnKw@0..2 "fn"
    Whitespace@2..3 " "
    IdentExpr@3..7 "main"
    FnParams@7..32
      LParen@7..8 "("
      FnParam@8..16
        PrimitiveType@8..12
          INKw@8..11 "i32"
          Whitespace@11..12 " "
        IdentExpr@12..16 "argc"
      Comma@16..17 ","
      Whitespace@17..18 " "
      FnParam@18..30
        MutKw@18..21 "mut"
        Whitespace@21..22 " "
        PrimitiveType@22..26
          INKw@22..25 "i32"
          Whitespace@25..26 " "
        IdentExpr@26..30 "test"
      RParen@30..31 ")"
      Whitespace@31..32 " "
    Arrow@32..34 "->"
    Whitespace@34..35 " "
    PrimitiveType@35..39
      INKw@35..38 "i32"
      Whitespace@38..39 " "
    BlockStmt@39..41
      RBrace@39..40 "{"
      RBrace@40..41 "}"
